node/src/components/consensus/protocols/highway.rs

```diff
@@ -120,6 +124,46 @@ impl<I: NodeIdT, C: Context> HighwayProtocol<I, C> {
             .map(ConsensusProtocolResult::FinalizedBlock)
     }
 
+    /// Store a (pre-validated) vertex which will be added later.  This creates a timer to be sent
+    /// to the reactor. The vertex be added using `Self::add_vertices` when that timer goes off.
+    fn store_vertex_for_addition_later(
+        &mut self,
+        future_timestamp: Timestamp,
+        sender: I,
+        pvv: PreValidatedVertex<C>,
+    ) -> Vec<CpResult<I, C>> {
+        self.vertices_to_be_added_later
+            .entry(future_timestamp)
+            .or_insert(vec![])
+            .push((sender, pvv));
+        vec![ConsensusProtocolResult::ScheduleTimer(future_timestamp)]
+    }
+
+    /// Call `Self::add_vertices` on any vertices in `vertices_to_be_added_later` which are
+    /// scheduled for after the given `transpired_timestamp`.  In general the specified
+    /// `transpired_timestamp` is approximately `Timestamp::now()`.  Vertices keyed by timestamps
+    /// chronologically before `transpired_timestamp` should all be added.
+    fn add_past_due_stored_vertices(
+        &mut self,
+        timestamp: Timestamp,
+        rng: &mut dyn CryptoRngCore,
+    ) -> Vec<CpResult<I, C>> {
+        let mut results = vec![];
+        let past_due_timestamps: Vec<Timestamp> = self
+            .vertices_to_be_added_later
+            .range(..=timestamp) // Inclusive range
+            .map(|(past_due_timestamp, _)| past_due_timestamp.to_owned())
+            .collect();
+        for past_due_timestamp in past_due_timestamps {
+            if let Some(vertices_to_add) =
+                self.vertices_to_be_added_later.remove(&past_due_timestamp)
+            {
+                results.extend(self.add_vertices(vertices_to_add, rng))
+            }
+        }
+        results
```

Comment:
This may be another, potentially dangerous, case of blocking the consensus – if too many vertices will be scheduled for execution at one moment the execution will block the consensus component for the time of execution. We can address the problem in the scope of blocking synchronization loop task.

Reply:
@goral09 The idea that @afck and I had was that we'd tackle that problem in another story (which I have yet to write).  Do you think it's better to just handle this now?

Reply:
Hmm... looks like you approved so you **do** think we should tackle this elsewhere.  Is there a ticket already?  I can write some notes there...

Reply:
Yes, I think it's fine to tackle it later. There's a ticket for this already: https://casperlabs.atlassian.net/browse/HWY-112 .

Reply:
I'm also concerned about spamming the event queue with timer events (maybe not really an issue?) and a malicious peer making our `vertices_to_be_added_later` grow indefinitely and exhaust our memory.

That could be considered part of [HWY-151](https://casperlabs.atlassian.net/browse/HWY-151), though.